core/src/main/java/io/grpc/internal/TransportSet.java

```diff
@@ -227,21 +227,23 @@ private Runnable startNewTransport(DelayedClientTransport delayedTransport) {
    * @param status the causal status when the channel begins transition to
    *     TRANSIENT_FAILURE.
    */
-  @CheckReturnValue
-  @GuardedBy("lock")
-  private Runnable scheduleBackoff(
+  private void scheduleBackoff(
       final DelayedClientTransport delayedTransport, final Status status) {
-    Preconditions.checkState(reconnectTask == null, "previous reconnectTask is not done");
+    // This must be run outside of lock. The TransportSet lock is a channel level lock.
+    // startBackoff() will acquire the delayed transport lock, which is a transport level
+    // lock. Our lock ordering mandates transport lock > channel lock.  Otherwise a deadlock
+    // could happen (https://github.com/grpc/grpc-java/issues/2152).
+    delayedTransport.startBackoff(status);
 
-    if (reconnectPolicy == null) {
-      reconnectPolicy = backoffPolicyProvider.get();
-    }
-    long delayMillis =
-        reconnectPolicy.nextBackoffMillis() - connectingTimer.elapsed(TimeUnit.MILLISECONDS);
-    if (log.isLoggable(Level.FINE)) {
-      log.log(Level.FINE, "[{0}] Scheduling backoff for {1} ms",
-          new Object[]{getLogId(), delayMillis});
+    synchronized (lock) {
+      if (!shutdown) {
+        scheduleEndOfCurrentBackoff(delayedTransport);
+      }
     }
+  }
+
+  @GuardedBy("lock")
+  private void scheduleEndOfCurrentBackoff(final DelayedClientTransport delayedTransport) {
```

Comment:
scheduleBackoff() only has 4 lines of code. It's by no means complicated.

> ... then one usually adds a comment before it for better readability like `// schedule end of current backoff`

That's fine, although I think the code is self-descriptive enough so it can live without additional comment.

>... which is easily becoming obsolete.

Why?